File: DailyChallenges/Scala/LeetCode_2316.scala
object Solution {
    def countPairs(n: Int, edges: Array[Array[Int]]): Long = {
        val p = (0 until n).toArray
        def find(i: Int): Int = {
            if (p(i) != i) p(i) = find(p(i))
            p(i)
        }

        for (e<-edges){
            val i1 = find(e(0))
            val i2 = find(e(1))
            if (i1 != i2) p(i1) = i2
        }

        val m = collection.mutable.Map.empty[Int, Int]
        for (i <- 0 until n; _i = find(i))
            m(_i) = m.getOrElse(_i, 0) + 1
        
        m.foldLeft((0L,n))((ap, kp) => (ap._1 + kp._2.toLong * (ap._2-kp._2), ap._2-kp._2))._1
    }
}

File: DailyChallenges/CPP/LeetCode_2316.cpp
class Solution {
public:
    long long countPairs(int n, vector<vector<int>>& edges) {
        vector<int> p(n);
        iota(p.begin(), p.end(), 0);
        
        function<int(int)> find = [&](int i) { return (p[i] == i)? p[i] : (p[i] = find(p[i])); };
    
        for (auto& e: edges){
            int i1 = find(e[0]);
            int i2 = find(e[1]);
            if (i1 != i2) p[i1] = i2;
        }

        unordered_map<int, int> m;
        for (int i=0; i<n; ++i)
            ++m[find(i)];
        
        long long ans = 0;
        for (auto [_, v]: m)
            ans += (n -= v) * (long long)v;

        return ans;
    }
};
